// @web3auth/default-evm-adapter/dist/defaultEvmAdapter.esm.js
import _objectSpread from '@babel/runtime/helpers/objectSpread2';
import { ADAPTER_NAMESPACES, CHAIN_NAMESPACES, ADAPTER_CATEGORY, ADAPTER_STATUS, ADAPTER_EVENTS, log, Web3AuthError, WalletLoginError, WalletInitializationError, getChainConfig, normalizeWalletName } from '@web3auth/base';
import { createStore } from 'mipd';
import _defineProperty from '@babel/runtime/helpers/defineProperty';
import { BaseEvmAdapter } from '@web3auth/base-evm-adapter';

class InjectedEvmAdapter extends BaseEvmAdapter {
  constructor(options) {
    super(options);
    _defineProperty(this, "adapterNamespace", ADAPTER_NAMESPACES.EIP155);
    _defineProperty(this, "currentChainNamespace", CHAIN_NAMESPACES.EIP155);
    _defineProperty(this, "type", ADAPTER_CATEGORY.EXTERNAL);
    _defineProperty(this, "name", void 0);
    _defineProperty(this, "status", ADAPTER_STATUS.NOT_READY);
    _defineProperty(this, "injectedProvider", null);
    this.name = options.name;
    this.injectedProvider = options.provider;
  }
  get provider() {
    if (this.status !== ADAPTER_STATUS.NOT_READY && this.injectedProvider) {
      return this.injectedProvider;
    }
    return null;
  }
  set provider(_) {
    throw new Error("Not implemented");
  }
  async init(options = {}) {
    await super.init(options);
    super.checkInitializationRequirements();
    this.status = ADAPTER_STATUS.READY;
    this.emit(ADAPTER_EVENTS.READY, this.name);
    try {
      log.debug(`initializing ${this.name} injected adapter`);
      if (options.autoConnect) {
        this.rehydrated = true;
        await this.connect();
      }
    } catch (error) {
      this.emit(ADAPTER_EVENTS.ERRORED, error);
    }
  }
  async connect() {
    super.checkConnectionRequirements();
    this.status = ADAPTER_STATUS.CONNECTING;
    this.emit(ADAPTER_EVENTS.CONNECTING, {
      adapter: this.name
    });
    try {
      var _this$injectedProvide;
      await ((_this$injectedProvide = this.injectedProvider) === null || _this$injectedProvide === void 0 ? void 0 : _this$injectedProvide.request({
        method: "eth_requestAccounts"
      }));
      this.status = ADAPTER_STATUS.CONNECTED;
      const disconnectHandler = () => {
        var _this$injectedProvide2;
        // ready to be connected again
        this.disconnect();
        (_this$injectedProvide2 = this.injectedProvider) === null || _this$injectedProvide2 === void 0 || _this$injectedProvide2.removeListener("disconnect", disconnectHandler);
      };
      this.injectedProvider.on("disconnect", disconnectHandler);
      this.emit(ADAPTER_EVENTS.CONNECTED, {
        adapter: this.name,
        reconnected: this.rehydrated,
        provider: this.injectedProvider
      });
      return this.injectedProvider;
    } catch (error) {
      // ready again to be connected
      this.status = ADAPTER_STATUS.READY;
      this.rehydrated = false;
      this.emit(ADAPTER_EVENTS.ERRORED, error);
      if (error instanceof Web3AuthError) throw error;
      throw WalletLoginError.connectionError(`Failed to login with ${this.name} injected wallet`);
    }
  }
  async disconnect(options = {
    cleanup: false
  }) {
    var _this$injectedProvide3, _this$injectedProvide4;
    await super.disconnectSession();
    if (typeof ((_this$injectedProvide3 = this.injectedProvider) === null || _this$injectedProvide3 === void 0 ? void 0 : _this$injectedProvide3.removeAllListeners) !== "undefined") (_this$injectedProvide4 = this.injectedProvider) === null || _this$injectedProvide4 === void 0 || _this$injectedProvide4.removeAllListeners();
    try {
      await this.injectedProvider.request({
        method: "wallet_revokePermissions",
        params: [{
          eth_accounts: {}
        }]
      });
    } catch (error) {}
    if (options.cleanup) {
      this.status = ADAPTER_STATUS.NOT_READY;
      this.injectedProvider = null;
    } else {
      // ready to be connected again
      this.status = ADAPTER_STATUS.READY;
    }
    await super.disconnect();
  }
  async getUserInfo() {
    if (this.status !== ADAPTER_STATUS.CONNECTED) throw WalletLoginError.notConnectedError("Not connected with wallet, Please login/connect first");
    return {};
  }
  async addChain(chainConfig, init = false) {
    var _this$injectedProvide5;
    super.checkAddChainRequirements(chainConfig, init);
    await ((_this$injectedProvide5 = this.injectedProvider) === null || _this$injectedProvide5 === void 0 ? void 0 : _this$injectedProvide5.request({
      method: "wallet_addEthereumChain",
      params: [{
        chainId: chainConfig.chainId,
        chainName: chainConfig.displayName,
        rpcUrls: [chainConfig.rpcTarget],
        blockExplorerUrls: [chainConfig.blockExplorerUrl],
        nativeCurrency: {
          name: chainConfig.tickerName,
          symbol: chainConfig.ticker,
          decimals: chainConfig.decimals || 18
        },
        iconUrls: [chainConfig.logo]
      }]
    }));
    this.addChainConfig(chainConfig);
  }
  async switchChain(params, init = false) {
    var _this$injectedProvide6;
    super.checkSwitchChainRequirements(params, init);
    await ((_this$injectedProvide6 = this.injectedProvider) === null || _this$injectedProvide6 === void 0 ? void 0 : _this$injectedProvide6.request({
      method: "wallet_switchEthereumChain",
      params: [{
        chainId: params.chainId
      }]
    }));
    this.setAdapterSettings({
      chainConfig: this.getChainConfig(params.chainId)
    });
  }
  async enableMFA() {
    throw new Error("Method Not implemented");
  }
}

const getDefaultExternalAdapters = async params => {
  const {
    options
  } = params;
  const {
    clientId,
    chainConfig,
    sessionTime,
    web3AuthNetwork,
    useCoreKitKey
  } = options;
  if (!Object.values(CHAIN_NAMESPACES).includes(chainConfig.chainNamespace)) throw WalletInitializationError.invalidParams(`Invalid chainNamespace: ${chainConfig.chainNamespace}`);
  const finalChainConfig = _objectSpread(_objectSpread({}, getChainConfig(chainConfig.chainNamespace, chainConfig === null || chainConfig === void 0 ? void 0 : chainConfig.chainId)), chainConfig || {});
  const [{
    TorusWalletAdapter
  }, {
    WalletConnectV2Adapter
  }] = await Promise.all([import('@web3auth/torus-evm-adapter'), import('@web3auth/wallet-connect-v2-adapter')]);
  const torusWalletAdapter = new TorusWalletAdapter({
    chainConfig: finalChainConfig,
    clientId,
    sessionTime,
    web3AuthNetwork,
    useCoreKitKey
  });
  const wcv2Adapter = new WalletConnectV2Adapter({
    chainConfig: finalChainConfig,
    clientId,
    sessionTime,
    web3AuthNetwork,
    useCoreKitKey,
    adapterSettings: {
      walletConnectInitOptions: {}
    }
  });

  // EIP-6963: multiple injected provider discovery
  const mipd = createStore();
  // We assume that all extensions have emitted by here.
  // TODO: Ideally, we must use reactive listening. We will do that with v9
  const injectedProviders = mipd.getProviders().map(providerDetail => {
    return new InjectedEvmAdapter({
      name: normalizeWalletName(providerDetail.info.name),
      provider: providerDetail.provider,
      chainConfig: finalChainConfig,
      clientId,
      sessionTime,
      web3AuthNetwork,
      useCoreKitKey
    });
  });
  return [...injectedProviders, torusWalletAdapter, wcv2Adapter];
};

export { getDefaultExternalAdapters };
